CurveXYZFourierSymmetries

andrewgiuliani opened the pull request:
This pull request introduces a new helical curve representation called `CurveXYZHelical`.  This curve is different from `CurveXYZFourier` because it can have discrete rotational symmetry.  It can also be stellarator symmetric, or not.

We currently support a helical curve representation `CurveHelical`, but this curve is restricted to lie on a torus.  The new  `CurveXYZHelical` is not restricted to lie on a torus, and can represent curves like the one below.  Viewing from above, the  `CurveXYZHelical` appears to lie on a torus
![helical_top](https://github.com/hiddenSymmetries/simsopt/assets/51761749/dd7524a8-7473-40f7-8679-89976ea5db3b)

but profile view reveals that it clearly does not:
![helical_side](https://github.com/hiddenSymmetries/simsopt/assets/51761749/ace19c13-414f-4e26-afbb-5e2633ef0a6f)

When `CurveXYZHelical` is stellarator symmetric, then necessarily the curve must pass through the point `(x0, 0, 0)`.  If you do not want this, then the user can apply stellarator symmetry to a non-stellarator symmetric `CurveXYZHelical` curve and obtain a stellsym magnetic field from those two curves.

VascoSch92 commented on src/simsopt/geo/curvexyzhelical.py:
```diff
@@ -0,0 +1,119 @@
+
+    def _make_names(self, order):
+        if self.stellsym:
+            x_cos_names = [f'xc({i})' for i in range(0, order + 1)]
+            x_names = x_cos_names
+            y_sin_names = [f'ys({i})' for i in range(1, order + 1)]
+            y_names = y_sin_names
+            z_sin_names = [f'zs({i})' for i in range(1, order + 1)]
+            z_names = z_sin_names
```
```suggestion
            x_names = [f'xc({i})' for i in range(1, order + 1)] # x cos names
            y_names = [f'ys({i})' for i in range(1, order + 1)] # y sin names
            z_names = [f'zs({i})' for i in range(1, order + 1)] # z sin names
```
moreover the range for x_names is starting from 0 instead of 1. Is it correct?

landreman reviewed (commented):
Here is a small generalization you could consider: in the transformation from `\hat{x},\hat{y}` to `x,y`, the angle in the rotation matrix could be multiplied by some integer. This would allow for curves for which you need to go around multiple times toroidally before they close. Not critical, just an idea.

andrewgiuliani commented on tests/geo/test_curve.py:
```diff
@@ -87,18 +95,152 @@ def get_curve(curvetype, rotated, x=np.asarray([0.5])):
+        curve1.set('xc(1)', r)
+        curve1.set('zs(1)', -r)
+        curve2 = CurveHelical(x, order, nfp, 1, R, r, x0=np.zeros((2*order,)))
+        assert np.mean(np.linalg.norm(curve1.gamma()-curve2.gamma(), axis=-1)) == 0 
```
done

andrewgiuliani commented:
> Here is a small generalization you could consider: in the transformation from `\hat{x},\hat{y}` to `x,y`, the angle in the rotation matrix could be multiplied by some integer. This would allow for curves for which you need to go around multiple times toroidally before they close. Not critical, just an idea.

sounds like a good idea, I'll add this to the PR in the coming days.

andrewgiuliani commented:
I've introduced support for curves to wrap toroidally `ntor` times before biting its tail.  This allows us to represent some more interesting curves with `nfp`-fold discrete rotational symmetry that pass `ntor` times toroidally before biting their tail.  In addition to helical coils, we can now represent knots such as a trefoil:

![trefoil](https://github.com/hiddenSymmetries/simsopt/assets/51761749/c04c907b-efcd-48b3-9559-094515250bca)

This trefoil has `nfp=3` and `ntor=2`.  I've added a unit test to verify this in `geo.test_curve.Testing.test_trefoil_stellsym` and  `geo.test_curve.Testing.test_trefoil_nonstellsym`.

A few things to note:
1. Since this is a much larger class of curves, i.e., not just helical curves, I've renamed the class to be `CurveXYZFourierSymmetries`, reflecting that the class has symmetries baked in (rotational and stellarator symmetries), that `CurveXYZFourier` does not.

2. Note that `ntor` and `nfp` must be coprime, or else the actual `nfp` of the curve is `nfp//gcd(nfp,ntor)` and the actual `ntor` of the curve is `ntor//gcd(nfp, ntor)`.  To avoid confusion, we assert that `gcd(nfp, ntor)=1` at instantiation of the class.

3. When `ntor=1` the usual relation showing rotational symmetry holds, i.e. `gamma(theta + 1/nfp) = R(1/nfp) * gamma(theta)`, where `R(alpha)` is a rotation matrix.  When `ntor>1`, a more general relation holds `gamma(theta + 1/nfp) = R(ntor/nfp) * gamma(theta)`.  This still results in a curve with `nfp`-fold discrete rotational symmetry.

rogeriojorge reviewed (approved):
Looks good to me!
